File: backend/src/main/java/org/example/backend/domain/point/entity/Point.java
package org.example.backend.domain.point.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.example.backend.domain.member.entity.Member;

import java.time.LocalDateTime;

@Entity
@Table(name = "point_log")
@Getter
@NoArgsConstructor
@AllArgsConstructor
public class Point {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long logId;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "member_id", nullable = false)
    private Member member;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private TransactionType type = TransactionType.CHARGE;

    @Column(nullable = false)
    private Long points = 0L;

    @Column(nullable = false)
    private Long afterPoint = 0L;

    @Column(nullable = false)
    private LocalDateTime createDate;

    // 내역 생성 - 충전
    public static Point create(Member member, long amount, long afterPoint) {
        Point point = new Point();
        point.member = member;
        point.type = TransactionType.CHARGE;
        point.points = amount;
        point.afterPoint = afterPoint;
        point.createDate = LocalDateTime.now();
        return point;
    }

    // 내역 생성 - 구매
    public static Point createPurchase(Member member, long amount, long afterPoint) {
        Point point = new Point();
        point.member = member;
        point.type = TransactionType.PURCHASE;
        point.points = amount;
        point.afterPoint = afterPoint;
        point.createDate = LocalDateTime.now();
        return point;
    }

    // 내역 생성 - 판매
    public static Point createSale(Member member, long amount, long afterPoint) {
        Point point = new Point();
        point.member = member;
        point.type = TransactionType.SALE;
        point.points = amount;
        point.afterPoint = afterPoint;
        point.createDate = LocalDateTime.now();
        return point;
    }
}

File: backend/src/main/java/org/example/backend/domain/point/entity/Point.kt
package org.example.backend.domain.point.entity

import jakarta.persistence.*
import org.example.backend.domain.member.entity.Member
import java.time.LocalDateTime

@Entity
@Table(name = "point_log")
class Point private constructor( // 해당 클래스만 접근 가능한 생성자
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "member_id", nullable = false)
    val member: Member,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    val type: TransactionType,

    @Column(nullable = false)
    val points: Long,

    @Column(nullable = false)
    val afterPoint: Long,

    @Column(nullable = false)
    val createDate: LocalDateTime = LocalDateTime.now()
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var logId: Long = 0L

    companion object {
        @JvmStatic
        fun create(member: Member, amount: Long, afterPoint: Long) = Point(
            member = member,
            type = TransactionType.CHARGE,
            points = amount,
            afterPoint = afterPoint
        )

        @JvmStatic
        fun createPurchase(member: Member, amount: Long, afterPoint: Long) = Point(
            member = member,
            type = TransactionType.PURCHASE,
            points = amount,
            afterPoint = afterPoint
        )

        @JvmStatic
        fun createSale(member: Member, amount: Long, afterPoint: Long) = Point(
            member = member,
            type = TransactionType.SALE,
            points = amount,
            afterPoint = afterPoint
        )
    }
}